---
title: "sWGA-WGS_report"
author: "EPPIcenter UCSF: Jared Honeycutt"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: true
    toc_float: true
  # runtime: shiny
---

```{r setup, include=FALSE}
# Knitr Options
knitr::opts_chunk$set(
  echo = FALSE,
	message = FALSE,
	warning = FALSE,
  fig.width = 9,
  fig.height = 6
)

library(knitr)
library(tidyverse)
library(plotly)
library(RColorBrewer)
library(data.table)
library(viridis)
library(DT)

# ## set working directory, only needed for Rstudio testing
# library(rstudioapi)
# setwd(dirname(getActiveDocumentContext()$path))

#library(shiny)
```

## Report overview

This report includes summary statistics for alignments performed on whole genome sequencing data (after sWGA) using the following command run on `r Sys.Date()`:

<code>nextflow run kn_partI.nf</code>

### Accompanying reports

For a Nextflow report on computing and task performance, see the accompanying file: 

<code> nextflow_report.html </code>

For evaluation of read quality and adapter trimming efficiency, see the file: 

<code> multiqc_report.html </code>

## Alignment metrics 

### *P.falciparum* vs *H.sapiens* mapped reads

Total mapped reads to the respective human and *Plasmodium* reference genomes within each sample.

```{r pf to hs read ratio, echo=FALSE, warning=FALSE}

pf_stats <-read_tsv("Bam_stat_summary_pf_final.tsv", show_col_types = FALSE)
colnames(pf_stats) <- gsub(",","",colnames(pf_stats))

hs_stats <-read_tsv("Bam_stat_summary_hs_final.tsv", show_col_types = FALSE)
colnames(hs_stats) <- gsub(",","", colnames(hs_stats))

pf_stats %>%
  select(sample_name, reads_mapped_pf) %>%
  left_join(hs_stats %>% select(sample_name, reads_mapped_hs), by = "sample_name") %>%
  mutate(pf_ratio = reads_mapped_pf/reads_mapped_hs) %>%
  pivot_longer(cols = !c("sample_name","pf_ratio"), values_to = "reads_mapped", names_to = "species") %>%
  mutate(species = case_when(str_detect(species, "hs") ~ "Human",
                             str_detect(species, "pf") ~ "P.falciparum")) %>%
  mutate(sample_name = str_split(sample_name, "_", n = 5)) %>%
  mutate(sample_name = map(.x = sample_name, .f = ~ pluck(.x, 2))) %>%
  mutate(sample_name = paste0(sample_name)) -> reads_long
  
reads_long %>%
  ggplot(aes(x = sample_name, y = reads_mapped, fill = species)) +
  geom_bar(stat = "identity", position = "dodge") +
  scale_fill_brewer(palette = "Paired") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60)) -> p1
  
ggplotly(p1, width = 800)

```

### Read ratios

Ratio of *P.falciparum* reads to human reads (y) plotted against *Pf* mapped reads (x, log scale).

```{r plot read ratio, echo = FALSE}

reads_long %>%
  filter(species == "P.falciparum") %>%
  ggplot(aes(x = log10(reads_mapped), y = pf_ratio, color = sample_name)) +
  geom_point() + 
  scale_color_viridis(discrete = T, direction = -1) +
  theme_minimal() +
  labs(title = "Pf:Human read ratio") -> p2

ggplotly(p2, width = 600)

```


### Insert size distributions

*P.falciparum* aligned fragment length counts, by sample.

Note to user: sequencing libraries containing fragments with shorter than intended insert sizes may influence downstream variant calling performance.

```{r insert sizes, include = FALSE}
insert_files <- list.files(pattern = "insert.txt", recursive = TRUE, full.names = TRUE)

insert_table <- data.table()

# structure the table
insert_table$insert_size <- NA
insert_table$All_Reads.fr_count <- NA
insert_table$filename <- NA

# read relevant file information
for (i in 1:length(insert_files)) {
  read.table(insert_files[i], skip = 10, header = TRUE) %>%
    mutate(filename = insert_files[i]) %>%
    select(insert_size, All_Reads.fr_count, filename) -> table_add
  
  rbind(insert_table, table_add, fill = TRUE) -> insert_table
}

## generate sample name column from file path

insert_table %>%
  filter(!is.na(insert_size)) %>% ## in order to pivot on this
  mutate(filesplit = str_split(filename, pattern = "/")) %>%
  mutate(sample = map(.x = filesplit, .f = ~pluck(.x[[3]]))) %>%
  mutate(sample = paste0(sample)) %>%
  mutate(sample = gsub(".insert.txt", "", sample)) -> insert_table
```

#### Insert sizes, raw counts

```{r plot raw counts of fragment lengths insert table, fig.width = 10, fig.height = 5, echo=FALSE}
insert_table %>%
  ggplot(aes(x = insert_size, y = All_Reads.fr_count, color = sample)) +
  geom_line() +
  labs(title = "<i>P.falciparum</i> aligned fragment length distribution", subtitle = "(raw counts)", y="fragment counts") +
  scale_color_viridis(discrete = TRUE, direction = -1) +
  xlim(0,500) +
  theme(legend.position = "right") -> p3

ggplotly(p3, width = 600)

```

#### Insert sizes, relative to total reads for that sample

*P.falciparum* insert size distributions, relative to total *Pf* aligned reads for that sample.

```{r plot relative insert size counts, echo=FALSE, warning=FALSE}
insert_table %>%
  group_by(sample) %>%
  summarize(total_fragments = sum(All_Reads.fr_count)) %>%
  ungroup() -> total_fragments

insert_table %>%
  left_join(total_fragments, by = "sample") %>%
  mutate(rel_reads = (All_Reads.fr_count/total_fragments)) -> insert_table

insert_table %>%
  ggplot(aes(x = insert_size, y = rel_reads, color = sample)) +
  geom_line() +
  labs(title = "Fragment length distribution, normalized to total reads within a sample", y = "Pf read counts (relative)") +
  scale_y_continuous(trans = "log10") +
  scale_color_viridis(discrete = T, direction = -1) +
  xlim(0,1000) +
  theme(legend.position = "right") -> p3

ggplotly(p3, width = 600)

```

## Bam file statistics: *P.falciparum*

### Plot bam statistics

#### Properly paired reads

```{r plot proper pairs p4, echo=FALSE}

pf_stats %>%
  ggplot(aes(reads_mapped_pf, reads_properly_paired_pf)) +
  geom_point(aes(color = sample_name), size = 2) +
  geom_abline(color = "grey", linetype = "dotdash") +
               labs(title = "Pf reads: mapped vs properly paired",
                    x = "Reads mapped",
                    y = "Reads properly paired") +
               scale_color_viridis(discrete = T, direction = -1)  +
               theme(legend.position = "none") -> p4

ggplotly(p4)
  

```

#### Reads unmapped

```{r plot unmapped p5, echo=FALSE}

pf_stats %>%
  ggplot(aes(reads_mapped_pf, reads_unmapped_pf)) +
  geom_point(aes(color = sample_name), size = 2) +
  geom_abline(color = "grey", linetype = "dotdash") +
               labs(title = "Pf reads: mapped vs unmapped",
                    x = "Reads mapped",
                    y = "Reads unmapped") +
               scale_color_viridis(discrete = T, direction = -1)  +
               theme(legend.position = "none") -> p5

ggplotly(p5)
  
```
#### Reads duplicated

```{r plot proper pairs p6, echo=FALSE}

pf_stats %>%
  ggplot(aes(reads_mapped_pf, reads_duplicated_pf)) +
  geom_point(aes(color = sample_name), size = 2) +
  geom_abline(color = "grey", linetype = "dotdash") +
               labs(title = "Pf reads: mapped vs duplicated",
                    x = "Reads mapped",
                    y = "Reads duplicated") +
               scale_color_viridis(discrete = T, direction = -1)  +
               theme(legend.position = "none") -> p6

ggplotly(p6)

```

Note: Check <code> multiqc_report.html </code> for additional measurements and plots of duplicated reads per sample.  

(Duplicate reads are those that start/end at the same alignment positions and are likely to be PCR duplicates that will be collapsed by <code> picardtools MarkDuplicates[Spark] </code> later on in the variant calling procedure).


```{r pf bamstat shiny, echo=FALSE}

# ui <- fluidPage(
#   titlePanel(title=h4("Pf bam file stats", align="center")),
#   sidebarPanel(
#     selectInput("var1", label = "plot x:",
#               choices = colnames(pf_stats)),
#     selectInput("var2", label = "plot y:",
#               choices = colnames(pf_stats))),
#   mainPanel(
#     # Show scatterplot
#     plotlyOutput(outputId = "scatterplot")
#     # Show datatable with package DT
#     # output$tbl(outputId = "")
#     )
# )
# 
# server1 <- function(input,output){
#   
#   output$scatterplot <- renderPlotly({
#     ggplotly(ggplot(pf_stats, aes_string(input$var1, input$var2)) +
#                geom_point(aes(color = sample_name), size = 2) +
#                #geom_smooth() +
#                geom_abline(color = "grey", linetype = "dotdash") +
#                labs(title = "Pf bam stat comparisons",
#                     x = input$var1,
#                     y = input$var2) +
#                scale_color_viridis(discrete = T, direction = -1)  +
#                theme(legend.position = "none")
#     )
#     
#   })
# }
# 
# shinyApp(ui, server1, options = list(height = 500))
# 
```

### *P.falciparum* bam statistics table

The above plots are based on collated statistics from each sample's *Pf* reads using the command:

<code> samtools stats \$sorted_pf_bam | grep ^SN | cut -f 2- | awk -F"\t" '{print \$2}' > ${pair_id}_bamstat_pf.tsv </code>

to yield the final summary file: <code>Bam_stat_summary_pf_final.tsv</code>

which is displayed below. Download here or access the file directly in the current directory of this report file if you would like to perform any further analysis.

Note: An equivalent file is generated for *Hs*-aligned reads: <code>Bam_stat_summary_hs_final.tsv</code> and can be found in the same directory as this report.

```{r bam stats table}

pf_stats %>%
  relocate(sample_name, .before = row_total_reads_pf) %>%
  datatable(extensions = 'Buttons',
            options = list(dom = 'Bfrtip', 
                           buttons = c('copy', 'csv'), 
                           scrollX = TRUE,
                           fixedColumns = list(leftColumns = 2)))

```





